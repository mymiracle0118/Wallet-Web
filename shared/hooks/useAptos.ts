import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
