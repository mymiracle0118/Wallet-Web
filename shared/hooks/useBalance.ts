import { useEffect, useMemo } from 'react'
import { useQuery } from '@tanstack/react-query'
import { BigNumber } from 'ethers'

import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

type BalanceResponse = {
  balance: BigNumber | number
  balanceFormatted: number
  isLoading: boolean
}
const { updateTokenBalance } = useStore.getState()
export const useBalance = (token: NetworkToken): BalanceResponse => {
  let balance, balanceFormatted

  const networkFactory = NetworkFactory.selectByNetworkId(token.shortName)
  if (!networkFactory) {
    return {
      balance: 0,
      balanceFormatted: 0,
      isLoading: false,
    }
  }

  const walletAddress = token.address
  const { data, isLoading } = useQuery(
    ['asset-balance', token.networkName, walletAddress, token.shortName],
    () => networkFactory.getTokenBalance(walletAddress as string, token),
    {
      refetchInterval: 60000,
      staleTime: 30000,
    }
  )

  balance = useMemo(() => {
    return data || BigNumber.from(0)
  }, [data])
  balanceFormatted = useMemo(() => networkFactory.getBalanceFormattedValue(data), [data]) as number

  useEffect(() => {
    // Update asset balance in wallet to keep running total
    if (balance && balanceFormatted) {
      updateTokenBalance(token.shortName, balance, balanceFormatted)
    }
  }, [balance, balanceFormatted])

  return {
    balance,
    balanceFormatted,
    isLoading,
  }
}
