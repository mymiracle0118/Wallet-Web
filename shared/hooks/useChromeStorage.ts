import { useCallback, useEffect, useState } from 'react'

export const storage = {
  get: (key: string, defaultValue: any, storageArea: Storage) => {
    const keyObj = defaultValue === undefined ? key : { [key]: defaultValue }
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].get(keyObj, (items: any) => {
        const error = chrome.runtime.lastError
        if (error) return reject(error)
        resolve(items[key])
      })
    })
  },
  set: (key: string, value: any, storageArea: Storage) => {
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].set({ [key]: value }, () => {
        const error = chrome.runtime.lastError
        error ? reject(error) : resolve()
      })
    })
  },
}

/**
 * Basic hook for storage
 * @param {string} key
 * @param {*} initialValue
 * @param {'local'|'sync'} storageArea
 * @returns {[*, function(*= any): void, boolean, string]}
 */

export default function useChromeStorage(key, initialValue, storageArea) {
  const [INITIAL_VALUE] = useState(() => {
    return typeof initialValue === 'function' ? initialValue() : initialValue
  })
  const [STORAGE_AREA] = useState(storageArea)
  const [state, setState] = useState(INITIAL_VALUE)
  const [isPersistent, setIsPersistent] = useState(true)
  const [error, setError] = useState('')

  useEffect(() => {
    storage
      .get(key, INITIAL_VALUE, STORAGE_AREA)
      .then((res) => {
        setState(res)
        setIsPersistent(true)
        setError('')
      })
      .catch((error) => {
        setIsPersistent(false)
        setError(error)
      })
  }, [key, INITIAL_VALUE, STORAGE_AREA])

  const updateValue = useCallback(
    (newValue) => {
      const toStore = typeof newValue === 'function' ? newValue(state) : newValue
      storage
        .set(key, toStore, STORAGE_AREA)
        .then(() => {
          setIsPersistent(true)
          setError('')
        })
        .catch((error) => {
          // set newValue to local state because chrome.storage.onChanged won't be fired in error case
          setState(toStore)
          setIsPersistent(false)
          setError(error)
        })
    },
    [STORAGE_AREA, key, state]
  )

  useEffect(() => {
    const onChange = (changes, areaName) => {
      if (areaName === STORAGE_AREA && key in changes) {
        setState(changes[key].newValue)
        setIsPersistent(true)
        setError('')
      }
    }
    chrome.storage.onChanged.addListener(onChange)
    return () => {
      chrome.storage.onChanged.removeListener(onChange)
    }
  }, [key, STORAGE_AREA])

  return [state, updateValue, isPersistent, error]
}
