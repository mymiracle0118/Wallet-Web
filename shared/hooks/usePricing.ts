import { useQuery } from '@tanstack/react-query'
import { utils } from 'ethers'
import React from 'react'

import { getSimplePrice } from '@portal/shared/services/coingecko'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

export const usePricing = (asset?: NetworkToken) => {
  const { currentTokenArrayWithBalance } = useStore()
  const { data } = useQuery(
    ['asset-pricing'],
    () =>
      getSimplePrice(
        [...new Set(Object.values(currentTokenArrayWithBalance).map((a) => a.coingeckoTokenId))].join(',')
      ),
    {
      refetchInterval: 40000,
      staleTime: 30000,
    }
  )

  const getMarketPrice = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        const total = Math.round((data[assetId].usd + Number.EPSILON) * 100) / 100
        return utils.commify(total)
      }

      return '0'
    },
    [data]
  )

  const getMarket24HourChange = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        return data[assetId].usd_24h_change
      }

      return 0
    },
    [data]
  )

  const getAssetValue = React.useCallback(
    (assetId: string, units: number) => {
      if (data && data[assetId]) {
        const value = Math.round((data[assetId].usd * units + Number.EPSILON) * 100) / 100
        return Math.max(value, 0)
      }

      return 0
    },
    [data]
  )

  const getAssetValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      return `$${getAssetValue(assetId, units)}`
    },
    [data]
  )
  const getAssetGasValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      let gasTotal = '0'
      if (data && data[assetId] && asset) {
        gasTotal = (data[assetId].usd * units).toFixed(parseInt(<string>asset?.decimal || '8'))
      }
      return `$${gasTotal}`
    },
    [data]
  )

  return { getMarketPrice, getMarket24HourChange, getAssetValue, getAssetValueFormatted, getAssetGasValueFormatted }
}
