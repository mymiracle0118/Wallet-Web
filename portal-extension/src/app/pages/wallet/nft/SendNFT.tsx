import React, { useState, useMemo, useEffect } from 'react'
import { useNavigate, createLocationState } from 'lib/woozie'
import { useTranslation } from 'react-i18next'
import { ethers } from 'ethers'

import { useWallet } from '@portal/shared/hooks/useWallet'
import { usePricing } from '@portal/shared/hooks/usePricing'
import { useGas } from '@portal/shared/hooks/useGas'
import { useSettings } from '@portal/shared/hooks/useSettings'
import type { AddressBookUser } from '@portal/shared/hooks/useSettings'
import { SAFETY_MEASURE } from 'utils/constants'

import SinglePageTitleLayout from 'layouts/single-page-layout/SinglePageLayout'
import { CustomTypography, Icon, AddressDropdown, Button } from 'components'

import { tokenImage } from 'utils/tokenImage'
import { GasOption } from '@portal/shared/utils/types'
import { Checkbox, Divider, Slider } from '@nextui-org/react'

const SendNFT = () => {
  const { t } = useTranslation()
  const { pathname } = createLocationState()
  const { navigate } = useNavigate()

  const paths = pathname.split('/')
  const network: string = paths[paths.length - 3]
  const contractAddress: string = paths[paths.length - 2]
  const assetId: string = paths[paths.length - 1]

  const { getNFTAsset, setTransaction } = useWallet()
  const asset = useMemo(
    () => getNFTAsset(network, contractAddress, assetId),
    [getNFTAsset, network, contractAddress, assetId]
  )

  const { getAssetValueFormatted } = usePricing()

  const {
    gasOption,
    setGasOption,
    maxFeePerGas,
    maxPriorityFeePerGas,
    estimatedTransactionCost,
    estimatedTransactionTime,
  } = useGas(network, asset.type as string, asset.contractAddress as string)

  const [selectedUser, setSelectedUser] = useState<null | AddressBookUser>(null)

  // handle address book
  const { addToAddressBook, addressBook } = useSettings()
  const [isAddressBookChecked, setIsAddressBookChecked] = useState<boolean>(false)
  useEffect(() => {
    setIsAddressBookChecked(false)
  }, [selectedUser])
  const addUserToAddressBook = () => {
    if (isAddressBookChecked && selectedUser) {
      addToAddressBook({
        avatar: '',
        username: `@TestAccount #${addressBook?.length + 1}`,
        address: selectedUser.address,
        safety: SAFETY_MEASURE.LOW,
      })
    }
  }

  const estimatedEtherFees = Number(ethers.utils.formatEther(estimatedTransactionCost as ethers.BigNumberish))

  const handleReviewTransaction = () => {
    if (selectedUser) {
      addUserToAddressBook()
      setTransaction({
        asset,
        to: selectedUser.address,
        gasOption,
      })
      navigate('/transaction/send-review')
    }
  }

  const onChangeAddress = (newValue: AddressBookUser) => {
    setSelectedUser(newValue)
  }

  const showAdd =
    !!selectedUser?.address.length &&
    !addressBook.filter((user: AddressBookUser) => user.address === selectedUser?.address).length

  return (
    <SinglePageTitleLayout showMenu>
      <div className="px-4">
        <div className="flex pb-4">
          <div className="h-8 w-8 text-xl">
            <Icon icon={tokenImage('ethereum')} size="inherit" />
          </div>
          <CustomTypography variant="h2" className="ml-4">
            {t('Nft.sendArtwork')}
          </CustomTypography>
        </div>
        <div className="gap-4">
          <img src={asset.metadata?.image} alt="nft-demo" className="rounded-lg min-w-[8.75rem]" />
          <div className="flex flex-col gap-1">
            <CustomTypography variant="h3">Description data is missing</CustomTypography>
            <CustomTypography variant="subtitle" type="secondary">
              Edition -- of {asset.id}
            </CustomTypography>
          </div>
        </div>
        <CustomTypography variant="subtitle" type="secondary" className="mb-3">
          {t('Labels.to')}
        </CustomTypography>
        <AddressDropdown
          selectedUser={selectedUser || undefined}
          data={addressBook}
          onChange={onChangeAddress}
          onBookmarkClick={() => navigate('/transaction/address-book')}
        />
        {showAdd && (
          <>
            <CustomTypography className="text-body my-2">Ethereum network</CustomTypography>
            <Checkbox onValueChange={setIsAddressBookChecked} isSelected={isAddressBookChecked}>
              Add to address book
            </Checkbox>
          </>
        )}
        <div className="mt-8 mb-1 flex">
          <div className="flex flex-1">
            <CustomTypography variant="subtitle" type="secondary">
              {t('Labels.maxFee')}
            </CustomTypography>
            <CustomTypography variant="subtitle" className="ml-1">
              {parseFloat(ethers.utils.formatUnits(maxFeePerGas as ethers.BigNumberish, 'gwei')).toFixed(2)} Gwei
            </CustomTypography>
          </div>
          <div className="flex flex-1">
            <CustomTypography variant="subtitle" type="secondary" className="flex-1 text-right">
              {t('Labels.maxPriorityFee')}
            </CustomTypography>
            <CustomTypography variant="subtitle" className="ml-1 min-w-[4.5em]">
              {parseFloat(ethers.utils.formatUnits(maxPriorityFeePerGas as ethers.BigNumberish, 'gwei'))} Gwei
            </CustomTypography>
          </div>
        </div>
        {/* <Slider
          marks
          aria-label="Volume"
          value={gasOption}
          onChange={(_, v: GasOption) => {
            setGasOption(v)
          }}
          color="secondary"
          size="small"
          min={1}
          max={3}
        /> */}
        <Slider
          size="sm"
          minValue={3}
          maxValue={1}
          aria-label="Volume"
          value={gasOption}
          onChange={(_, v: GasOption) => {
            setGasOption(v)
          }}
        />
        F
        <div className="mt-3">
          <CustomTypography variant="subtitle" className="flex-1">
            {t('Token.estimatedGasFee')}
          </CustomTypography>
          <div>
            <CustomTypography className="text-right" variant="subtitle">
              {estimatedEtherFees.toFixed(8)} ETH
            </CustomTypography>
            <CustomTypography variant="subtitle" className="text-right">
              {getAssetValueFormatted('ethereum', estimatedEtherFees)}
            </CustomTypography>
          </div>
        </div>
        <Divider className="mt-2" />
        <div className="mt-4">
          <CustomTypography variant="subtitle" className="flex-1">
            {t('Token.estimatedTime')}
          </CustomTypography>
          <div>
            <CustomTypography className="text-right" variant="subtitle">
              {estimatedTransactionTime}
            </CustomTypography>
          </div>
        </div>
        <div className="mt-4 flex flex-col justify-between items-center">
          <Button
            data-test-id="button-review"
            onClick={handleReviewTransaction}
            color={`${!ethers.utils.isAddress(selectedUser?.address as string) ? 'disabled' : 'primary'}`}
            isDisabled={!ethers.utils.isAddress(selectedUser?.address as string)}
          >
            {t('Token.reviewTransaction')}
          </Button>
          <Button variant="bordered" color="outlined" isDisabled className="mt-1 mb-2">
            {t('Token.sendWithLowGas')}
          </Button>
        </div>
      </div>
    </SinglePageTitleLayout>
  )
}

export default SendNFT
