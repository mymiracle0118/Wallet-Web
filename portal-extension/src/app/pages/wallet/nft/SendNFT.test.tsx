import React from 'react'
import { render, renderHook, act } from '@testing-library/react'
import { I18nextProvider } from 'react-i18next'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

import { useWallet } from '@portal/shared/hooks/useWallet'
import i18n from '@portal/shared/i18n/config'

import SendNFT from './SendNFT'

const queryClient = new QueryClient()
const wrapper = (
  <QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
      <SendNFT />
    </I18nextProvider>
  </QueryClientProvider>
)

describe('<SendNFT />', () => {
  const originalLocation = window.location as URL
  const mockWindowLocation = (newLocation: URL) => {
    delete window.location
    window.location = newLocation
  }
  const setLocation = (path: string) => mockWindowLocation(new URL(`https://example.com${path}`))

  beforeEach(() => {
    const { result } = renderHook(() => useWallet())

    act(() => {
      result.current.address = '0x63C100ac0C36549C7218070294b60C18d813675c'
      result.current.NFTs = [
        {
          id: '1',
          name: 'test',
          image: 'test',
          network: 'goerli',
          assetType: 'ERC721',
          chain: 'ethereum',
          contractAddress: '0x7a0c1749b3f55d0be14f38c406fbcae08094bf2a3079e1c459a819d1d57d916e',
        },
      ]
      setLocation('#/nft/goerli/0x7a0c1749b3f55d0be14f38c406fbcae08094bf2a3079e1c459a819d1d57d916e/1')
    })
  })

  afterEach(() => {
    // Restore window.location to not destroy other tests
    mockWindowLocation(originalLocation)
  })

  test('renders correctly', () => {
    const tree = render(wrapper)
    expect(tree).toMatchSnapshot()
  })
})
