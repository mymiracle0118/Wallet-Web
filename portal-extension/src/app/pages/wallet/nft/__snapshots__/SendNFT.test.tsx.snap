// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SendNFT /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
      >
        <div
          class="flex flex-row relative justify-between items-center h-16"
        >
          <button
            class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center"
            type="button"
          >
            <arrow-left.svg
              classname="text-xl stroke-custom-black dark:stroke-custom-white"
            />
          </button>
          <button
            class="profile-btn flex items-center mr-1 text-inherit font-bold hover:bg-custom-grey10 dark:hover:bg-custom-white10 p-3 rounded-md"
            type="button"
          >
            @
            <chevron-right.svg
              classname="stroke-custom-black dark:stroke-custom-white"
            />
          </button>
        </div>
        <div
          class="w-full h-full overflow-auto"
        >
          <div
            class="MuiBox-root css-10cpdan"
          >
            <div
              class="MuiBox-root css-xgi4bw"
            >
              <div
                class="MuiBox-root css-hxa509"
              >
                <svg
                  class="inline-block select-none shrink-0 fill-custom-black dark:fill-custom-white text-[inherit]]"
                  style="height: 1em; width: 1em;"
                >
                  <eth.svg />
                </svg>
              </div>
              <div
                class="text-custom-black dark:text-custom-white text-[1.75rem] font-light"
                style="margin-left: 16px;"
              >
                Send Artwork
              </div>
            </div>
            <div
              class="MuiBox-root css-13t3iqc"
            >
              <img
                alt="nft-demo"
                style="border-radius: 12px; min-width: 140px;"
              />
              <div
                class="MuiBox-root css-1u72um6"
              >
                <div
                  class="text-custom-black dark:text-custom-white text-lg font-extrabold"
                >
                  Description data is missing
                </div>
                <div
                  class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
                >
                  Edition -- of 
                  1
                </div>
              </div>
            </div>
            <div
              class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
              style="margin-bottom: 8px;"
            >
              To
            </div>
            <div
              class="border-solid  dark:border-custom-white40 border-custom-grey40 rounded-md"
              style="border-width: 0px;"
            >
              <div
                class="relative"
              >
                <div
                  class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
                >
                  <input
                    autocomplete="off"
                    class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                    id="headlessui-popover-button-:r0:"
                    placeholder="Username or address"
                    value=""
                  />
                  <div
                    class="pr-3.5"
                  >
                    <div
                      class="css-171n8m5"
                    >
                      <book.svg />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="mt-8 mb-1 MuiBox-root css-k008qs"
            >
              <div
                class="flex-1 MuiBox-root css-k008qs"
              >
                <div
                  class="text-custom-grey100 dark:text-custom-white40"
                >
                  Max Fee
                </div>
                <div
                  class="ml-1 text-custom-black dark:text-custom-white"
                >
                  0.00
                   Gwei
                </div>
              </div>
              <div
                class="flex-1 MuiBox-root css-k008qs"
              >
                <div
                  class="flex-1 text-right text-custom-grey100 dark:text-custom-white40"
                >
                  Max Priority Fee
                </div>
                <div
                  class="ml-1 min-w-[4.5em] text-custom-black dark:text-custom-white"
                >
                  0
                   Gwei
                </div>
              </div>
            </div>
            <span
              class="MuiSlider-colorSecondary MuiSlider-sizeSmall MuiSlider-root css-lbpym0-MuiSlider-root"
            >
              <span
                class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
              />
              <span
                class="MuiSlider-track css-1n40zqk-MuiSlider-track"
                style="left: 0%; width: 50%;"
              />
              <span
                class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
                data-index="0"
                style="left: 0%;"
              />
              <span
                class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
                data-index="1"
                style="left: 50%;"
              />
              <span
                class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
                data-index="2"
                style="left: 100%;"
              />
              <span
                class="MuiSlider-thumbColorSecondary MuiSlider-thumbSizeSmall MuiSlider-thumb css-nfrlyi-MuiSlider-thumb"
                data-index="0"
                style="left: 50%;"
              >
                <input
                  aria-label="Volume"
                  aria-orientation="horizontal"
                  aria-valuemax="3"
                  aria-valuemin="1"
                  aria-valuenow="2"
                  data-index="0"
                  max="3"
                  min="1"
                  step="1"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="2"
                />
              </span>
            </span>
            <div
              class="MuiBox-root css-1hprwxc"
            >
              <div
                class="text-custom-black dark:text-custom-white text-sm font-extrabold"
                style="flex: 1;"
              >
                Estimated Gas Fee
              </div>
              <div>
                <div
                  class="text-right text-custom-black dark:text-custom-white text-sm font-extrabold"
                >
                  0.00000000
                   ETH
                </div>
                <div
                  class="text-right text-custom-black dark:text-custom-white"
                >
                  $0
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-1h3sccu-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-mifb2i"
            >
              <div
                class="text-custom-black dark:text-custom-white text-sm font-extrabold"
                style="flex: 1;"
              >
                Estimated Time
              </div>
              <div>
                <div
                  class="text-custom-black dark:text-custom-white text-sm font-extrabold"
                  style="text-align: right;"
                >
                  30 seconds
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root css-r0pdzr"
            >
              <button
                class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
                data-testid="button-review"
                disabled=""
                type="button"
              >
                Review Transaction
              </button>
              <button
                class="mt-1 mb-2 flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-grey40 dark:border-grey40 hover:bg-transparent dark:hover:bg-transparent bg-none"
                disabled=""
                type="button"
              >
                Send with lower gas fee later
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
    >
      <div
        class="flex flex-row relative justify-between items-center h-16"
      >
        <button
          class="h-[24px] w-[24px] p-1 ml-1 hover:bg-custom-grey10 dark:hover:bg-custom-white10 rounded-full flex items-center justify-center"
          type="button"
        >
          <arrow-left.svg
            classname="text-xl stroke-custom-black dark:stroke-custom-white"
          />
        </button>
        <button
          class="profile-btn flex items-center mr-1 text-inherit font-bold hover:bg-custom-grey10 dark:hover:bg-custom-white10 p-3 rounded-md"
          type="button"
        >
          @
          <chevron-right.svg
            classname="stroke-custom-black dark:stroke-custom-white"
          />
        </button>
      </div>
      <div
        class="w-full h-full overflow-auto"
      >
        <div
          class="MuiBox-root css-10cpdan"
        >
          <div
            class="MuiBox-root css-xgi4bw"
          >
            <div
              class="MuiBox-root css-hxa509"
            >
              <svg
                class="inline-block select-none shrink-0 fill-custom-black dark:fill-custom-white text-[inherit]]"
                style="height: 1em; width: 1em;"
              >
                <eth.svg />
              </svg>
            </div>
            <div
              class="text-custom-black dark:text-custom-white text-[1.75rem] font-light"
              style="margin-left: 16px;"
            >
              Send Artwork
            </div>
          </div>
          <div
            class="MuiBox-root css-13t3iqc"
          >
            <img
              alt="nft-demo"
              style="border-radius: 12px; min-width: 140px;"
            />
            <div
              class="MuiBox-root css-1u72um6"
            >
              <div
                class="text-custom-black dark:text-custom-white text-lg font-extrabold"
              >
                Description data is missing
              </div>
              <div
                class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
              >
                Edition -- of 
                1
              </div>
            </div>
          </div>
          <div
            class="text-custom-grey100 dark:text-custom-white40 text-sm font-extrabold"
            style="margin-bottom: 8px;"
          >
            To
          </div>
          <div
            class="border-solid  dark:border-custom-white40 border-custom-grey40 rounded-md"
            style="border-width: 0px;"
          >
            <div
              class="relative"
            >
              <div
                class="flex relative rounded-md items-center hover:shadow-md w-full bg-input-light-bg dark:bg-custom-white10"
              >
                <input
                  autocomplete="off"
                  class="outline-none	py-3.5 px-3.5 w-full text-sm text-custom-black dark:text-custom-white bg-transparent bg-transparent text-h4 "
                  id="headlessui-popover-button-:r0:"
                  placeholder="Username or address"
                  value=""
                />
                <div
                  class="pr-3.5"
                >
                  <div
                    class="css-171n8m5"
                  >
                    <book.svg />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mt-8 mb-1 MuiBox-root css-k008qs"
          >
            <div
              class="flex-1 MuiBox-root css-k008qs"
            >
              <div
                class="text-custom-grey100 dark:text-custom-white40"
              >
                Max Fee
              </div>
              <div
                class="ml-1 text-custom-black dark:text-custom-white"
              >
                0.00
                 Gwei
              </div>
            </div>
            <div
              class="flex-1 MuiBox-root css-k008qs"
            >
              <div
                class="flex-1 text-right text-custom-grey100 dark:text-custom-white40"
              >
                Max Priority Fee
              </div>
              <div
                class="ml-1 min-w-[4.5em] text-custom-black dark:text-custom-white"
              >
                0
                 Gwei
              </div>
            </div>
          </div>
          <span
            class="MuiSlider-colorSecondary MuiSlider-sizeSmall MuiSlider-root css-lbpym0-MuiSlider-root"
          >
            <span
              class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
            />
            <span
              class="MuiSlider-track css-1n40zqk-MuiSlider-track"
              style="left: 0%; width: 50%;"
            />
            <span
              class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
              data-index="0"
              style="left: 0%;"
            />
            <span
              class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
              data-index="1"
              style="left: 50%;"
            />
            <span
              class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
              data-index="2"
              style="left: 100%;"
            />
            <span
              class="MuiSlider-thumbColorSecondary MuiSlider-thumbSizeSmall MuiSlider-thumb css-nfrlyi-MuiSlider-thumb"
              data-index="0"
              style="left: 50%;"
            >
              <input
                aria-label="Volume"
                aria-orientation="horizontal"
                aria-valuemax="3"
                aria-valuemin="1"
                aria-valuenow="2"
                data-index="0"
                max="3"
                min="1"
                step="1"
                style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                type="range"
                value="2"
              />
            </span>
          </span>
          <div
            class="MuiBox-root css-1hprwxc"
          >
            <div
              class="text-custom-black dark:text-custom-white text-sm font-extrabold"
              style="flex: 1;"
            >
              Estimated Gas Fee
            </div>
            <div>
              <div
                class="text-right text-custom-black dark:text-custom-white text-sm font-extrabold"
              >
                0.00000000
                 ETH
              </div>
              <div
                class="text-right text-custom-black dark:text-custom-white"
              >
                $0
              </div>
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-1h3sccu-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-mifb2i"
          >
            <div
              class="text-custom-black dark:text-custom-white text-sm font-extrabold"
              style="flex: 1;"
            >
              Estimated Time
            </div>
            <div>
              <div
                class="text-custom-black dark:text-custom-white text-sm font-extrabold"
                style="text-align: right;"
              >
                30 seconds
              </div>
            </div>
          </div>
          <div
            class="MuiBox-root css-r0pdzr"
          >
            <button
              class="flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] text-custom-white dark:text-custom-white bg-custom-grey10"
              data-testid="button-review"
              disabled=""
              type="button"
            >
              Review Transaction
            </button>
            <button
              class="mt-1 mb-2 flex items-center justify-center rounded-md hover:cursor-pointer font-extrabold text-sm w-full disabled:text-custom-grey40 dark:disabled:text-custom-grey40 h-[44px] border-[1px] border-solid text-primary dark:text-custom-white border-grey40 dark:border-grey40 hover:bg-transparent dark:hover:bg-transparent bg-none"
              disabled=""
              type="button"
            >
              Send with lower gas fee later
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
